egs/swbd/s5c/local/chain/run_discriminative.sh

```diff
@@ -0,0 +1,188 @@
+#!/bin/bash
+
+set -e 
+set -o pipefail
+
+# this is run_discriminative.sh
+
+# This script does discriminative training on top of chain nnet3 system.
+# note: this relies on having a cluster that has plenty of CPUs as well as GPUs,
+# since the lattice generation runs in about real-time, so takes of the order of
+# 1000 hours of CPU time.
+# 
+. cmd.sh
+
+
+stage=0
+train_stage=-10 # can be used to start training in the middle.
+get_egs_stage=-10
+use_gpu=true  # for training
+cleanup=false  # run with --cleanup true --stage 6 to clean up (remove large things like denlats,
+               # alignments and degs).
+
+set -e
+. cmd.sh
+. ./path.sh
+. ./utils/parse_options.sh
+
+srcdir=exp/nnet3/nnet_ms_a
+train_data_dir=data/train_nodup_sp_hires
+online_ivector_dir=exp/nnet3/ivectors_train_nodup_sp
+degs_dir=
+lats_dir=
+
+## Objective options
+criterion=smbr
+one_silence_class=true
+
+dir=${srcdir}_${criterion}
+
+## Egs options
+frames_per_eg=150
+frames_overlap_per_eg=30
+truncate_deriv_weights=10
+
+## Nnet training options
+effective_learning_rate=0.0000125
+max_param_change=1
+num_jobs_nnet=4
+num_epochs=4
+regularization_opts=
+minibatch_size=64
+adjust_priors=true
+modify_learning_rates=true
+last_layer_factor=1.0
+
+## Decode options
+decode_start_epoch=1 # can be used to avoid decoding all epochs, e.g. if we decided to run more.
+
+if $use_gpu; then
+  if ! cuda-compiled; then
+    cat <<EOF && exit 1 
+This script is intended to be used with GPUs but you have not compiled Kaldi with CUDA 
+If you want to use GPUs (and have them), go to src/, and configure and make on a machine
+where "nvcc" is installed.  Otherwise, call this script with --use-gpu false
+EOF
+  fi
+  num_threads=1
+else
+  # Use 4 nnet jobs just like run_4d_gpu.sh so the results should be
+  # almost the same, but this may be a little bit slow.
+  num_threads=16
+fi
+
+if [ ! -f ${srcdir}/final.mdl ]; then
+  echo "$0: expected ${srcdir}/final.mdl to exist; first run run_tdnn.sh or run_lstm.sh"
+  exit 1;
+fi
+
+lang=data/lang
+
+if [ $stage -le 1 ]; then
+  # hardcode no-GPU for alignment, although you could use GPU [you wouldn't
+  # get excellent GPU utilization though.]
+  nj=350 # have a high number of jobs because this could take a while, and we might
+         # have some stragglers.
+  use_gpu=no
+
+  steps/nnet3/align.sh  --cmd "$decode_cmd" --use-gpu "$use_gpu" \
+     --online-ivector-dir $online_ivector_dir \
+     --scale-opts "--transition-scale=1.0 --acoustic-scale=1.0 --self-loop-scale=1.0" \
+     --nj $nj $train_data_dir $lang $srcdir ${srcdir}_ali || exit 1;
```

Comment:
Ok. Will change.
There is also an issue of lstm vs tdnn models. Lstm models need
extra-left-context and extra-right-context. Can these be taken in as
parameters from the user?

On Thu, Mar 10, 2016 at 3:42 PM Daniel Povey <notifications@github.com>
wrote:

> In egs/swbd/s5c/local/chain/run_discriminative.sh
> <https://github.com/kaldi-asr/kaldi/pull/574#discussion_r55746123>:
>
> > +  exit 1;
> > +fi
> > +
> > +lang=data/lang
> > +
> > +if [ $stage -le 1 ]; then
> > +  # hardcode no-GPU for alignment, although you could use GPU [you wouldn't
> > +  # get excellent GPU utilization though.]
> > +  nj=350 # have a high number of jobs because this could take a while, and we might
> > +         # have some stragglers.
> > +  use_gpu=no
> > +
> > +  steps/nnet3/align.sh  --cmd "$decode_cmd" --use-gpu "$use_gpu" \
> > +     --online-ivector-dir $online_ivector_dir \
> > +     --scale-opts "--transition-scale=1.0 --acoustic-scale=1.0 --self-loop-scale=1.0" \
> > +     --nj $nj $train_data_dir $lang $srcdir ${srcdir}_ali || exit 1;
>
> in more recent scripts, we use set -e more and more. I think having set -e
> is fine.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/574/files#r55746123>.
>
-- 
Vimal Manohar
PhD Student
Electrical & Computer Engineering
Johns Hopkins University